sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveHints.scala

```diff
@@ -102,6 +104,35 @@ object ResolveHints {
     }
   }
 
+  /**
+   * COALESCE Hint accepts name "COALESCE" and "REPARTITION".
+   * Its parameter includes a partition number.
+   */
+  class ResolveCoalesceHints(conf: SQLConf) extends Rule[LogicalPlan] {
+    private val COALESCE_HINT_NAMES = Set("COALESCE", "REPARTITION")
+
+    private def applyCoalesceHint(
+      plan: LogicalPlan,
+      numPartitions: Int,
+      shuffle: Boolean): LogicalPlan = {
+      Repartition(numPartitions, shuffle, plan)
+    }
+
+    def apply(plan: LogicalPlan): LogicalPlan = plan.resolveOperators {
+      case h: UnresolvedHint if COALESCE_HINT_NAMES.contains(h.name.toUpperCase(Locale.ROOT)) =>
+        h.parameters match {
+          case Seq(Literal(numPartitions: Int, IntegerType)) =>
+            val shuffle = h.name.toUpperCase(Locale.ROOT) match {
+              case "REPARTITION" => true
+              case "COALESCE" => false
+            }
+            applyCoalesceHint(h.child, numPartitions, shuffle)
+          case _ =>
+            throw new AnalysisException("COALESCE Hint expects a partition number as parameter")
```

Comment:
Can you use `h.name.toUpperCase` in this error message instead? I think that would be a better message for users that don't know the relationship between COALESCE and REPARTITION.

Reply:
Good catch